File: ml-service/app/models/predict_request.py
from datetime import date

from pydantic import BaseModel, Field


class VendaDiaria(BaseModel):
    """Registro de venda de um único dia."""

    data: date
    quantidade: int = Field(gt=0)


class PredictRequest(BaseModel):
    """Payload de entrada para o endpoint POST /predict."""

    produto_id: int
    historico: list[VendaDiaria] = Field(min_length=90)
    lead_time_medio: int = Field(default=3, ge=1)
    variabilidade_lead_time: float = Field(default=1.0, ge=0)
    nivel_servico_alvo: float = Field(default=0.95, ge=0.5, le=0.999)
    estoque_atual: int = Field(ge=0)
    is_promocional: list[int] = Field(default_factory=list)

File: backend/src/main/kotlin/br/com/stocksense/client/dto/PredictRequest.kt
package br.com.stocksense.client.dto

import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.databind.annotation.JsonNaming
import java.time.LocalDate

/**
 * Contrato de entrada do `POST /predict` do ml-service. Espelha o Pydantic
 * `PredictRequest`. O ml-service usa snake_case no JSON — `@JsonNaming` faz a
 * ponte a partir do camelCase do Kotlin, sem alterar o ObjectMapper global.
 */
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
data class PredictRequest(
    val produtoId: Int,
    val historico: List<VendaDiaria>,
    val leadTimeMedio: Int,
    val variabilidadeLeadTime: Double,
    val nivelServicoAlvo: Double,
    val estoqueAtual: Int,
    val isPromocional: List<Int> = emptyList(),
)

@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
data class VendaDiaria(
    val data: LocalDate,
    val quantidade: Int,
)
